File: platform/android-seller/app/src/main/java/ca/shawcliffe/seller/InquiriesService.kt
package ca.shawcliffe.seller

import io.github.jan.supabase.auth.auth
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.engine.okhttp.OkHttp
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.client.request.parameter
import io.ktor.client.request.patch
import io.ktor.client.request.setBody
import io.ktor.client.statement.HttpResponse
import io.ktor.http.ContentType
import io.ktor.http.contentType
import io.ktor.http.isSuccess
import io.ktor.serialization.kotlinx.json.json
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

// Reads and updates this client's inquiries from the web backend, gated
// server-side on the inquiry_quote_form component (same Bearer-token pattern
// as CustomersService).
object InquiriesService {
    @Serializable
    data class InquiryCustomer(
        val name: String,
        val phone: String? = null,
        val email: String? = null,
    )

    @Serializable
    data class InquiryVehicle(
        val make: String? = null,
        val model: String? = null,
        val year: Int? = null,
        val vin: String? = null,
        val plate: String? = null,
        val mileage: Int? = null,
    )

    @Serializable
    data class InquiryProperty(
        val address: String? = null,
        @SerialName("gate_code") val gateCode: String? = null,
        @SerialName("parking_instructions") val parkingInstructions: String? = null,
        @SerialName("pets_on_site") val petsOnSite: String? = null,
        @SerialName("access_notes") val accessNotes: String? = null,
        @SerialName("preferred_service_day") val preferredServiceDay: String? = null,
        @SerialName("lawn_size") val lawnSize: String? = null,
        @SerialName("snow_removal_areas") val snowRemovalAreas: String? = null,
        @SerialName("cleaning_instructions") val cleaningInstructions: String? = null,
        @SerialName("safety_notes") val safetyNotes: String? = null,
    )

    @Serializable
    data class Inquiry(
        val id: String,
        @SerialName("service_category") val serviceCategory: String? = null,
        @SerialName("job_location") val jobLocation: String? = null,
        val urgency: String? = null,
        val description: String? = null,
        @SerialName("photo_urls") val photoUrls: List<String> = emptyList(),
        @SerialName("preferred_contact_method") val preferredContactMethod: String? = null,
        val status: String,
        @SerialName("created_at") val createdAt: String,
        val customer: InquiryCustomer? = null,
        val vehicle: InquiryVehicle? = null,
        val property: InquiryProperty? = null,
    )

    @Serializable
    data class Response(
        val inquiries: List<Inquiry> = emptyList(),
        val total: Int = 0,
        val page: Int = 0,
        val pageSize: Int = 25,
    )

    @Serializable
    private data class StatusUpdate(val status: String)

    private val client = HttpClient(OkHttp) {
        install(ContentNegotiation) { json(Json { ignoreUnknownKeys = true }) }
    }

    suspend fun fetch(status: String? = null, page: Int = 0): Response {
        val token = supabase.auth.currentAccessTokenOrNull() ?: throw BroadcastError.NotAuthenticated

        val response: HttpResponse = client.get("${Config.API_BASE_URL}/api/seller/inquiries") {
            header("Authorization", "Bearer $token")
            parameter("page", page)
            if (!status.isNullOrBlank()) parameter("status", status)
        }
        if (!response.status.isSuccess()) {
            throw BroadcastError.Server("Couldn't load inquiries.")
        }
        return response.body<Response>()
    }

    suspend fun updateStatus(inquiryId: String, status: String) {
        val token = supabase.auth.currentAccessTokenOrNull() ?: throw BroadcastError.NotAuthenticated

        val response: HttpResponse = client.patch("${Config.API_BASE_URL}/api/seller/inquiries/$inquiryId") {
            header("Authorization", "Bearer $token")
            contentType(ContentType.Application.Json)
            setBody(StatusUpdate(status))
        }
        if (!response.status.isSuccess()) {
            throw BroadcastError.Server("Couldn't update inquiry.")
        }
    }
}

File: clients/little-dough-bread-co/ios-seller/ShawcliffeSeller/InquiriesService.swift
import Foundation

// Reads/updates the current client's inquiries from the web backend. Mirrors
// CustomersService's Bearer-token GET pattern against /api/seller/inquiries,
// plus a PATCH for status updates (error handling mirrors BroadcastService).
enum InquiriesService {
    struct Response: Decodable {
        let inquiries: [Inquiry]
        let total: Int
        let page: Int
        let pageSize: Int
    }

    private struct StatusUpdate: Encodable {
        let status: String
    }

    private struct ErrorBody: Decodable {
        let error: String?
    }

    static func fetch(status: String? = nil, page: Int = 0) async throws -> Response {
        guard let token = try? await supabase.auth.session.accessToken else {
            throw BroadcastError.notAuthenticated
        }

        var components = URLComponents(
            url: Config.apiBaseURL.appendingPathComponent("api/seller/inquiries"),
            resolvingAgainstBaseURL: false
        )!
        var queryItems = [URLQueryItem(name: "page", value: String(page))]
        if let status, !status.isEmpty {
            queryItems.append(URLQueryItem(name: "status", value: status))
        }
        components.queryItems = queryItems

        var request = URLRequest(url: components.url!)
        request.httpMethod = "GET"
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw BroadcastError.server("Couldn't load inquiries.")
        }
        return try JSONDecoder().decode(Response.self, from: data)
    }

    static func updateStatus(inquiryId: String, status: String) async throws {
        guard let token = try? await supabase.auth.session.accessToken else {
            throw BroadcastError.notAuthenticated
        }

        var request = URLRequest(
            url: Config.apiBaseURL.appendingPathComponent("api/seller/inquiries/\(inquiryId)")
        )
        request.httpMethod = "PATCH"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        request.httpBody = try JSONEncoder().encode(StatusUpdate(status: status))

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse else {
            throw BroadcastError.server("No response from server.")
        }
        guard (200..<300).contains(http.statusCode) else {
            let message = (try? JSONDecoder().decode(ErrorBody.self, from: data))?.error
                ?? "Couldn't update inquiry (\(http.statusCode))."
            throw BroadcastError.server(message)
        }
    }
}
